File: src/models/Leave.js
const mongoose = require('mongoose');

const leaveSchema = new mongoose.Schema({
  leaveId: {
    type: String,
    unique: true,
    required: true
  },
  student: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: [true, 'Student is required']
  },
  leaveType: {
    type: String,
    enum: ['home', 'medical', 'emergency', 'personal', 'academic', 'other'],
    required: [true, 'Leave type is required']
  },
  startDate: {
    type: Date,
    required: [true, 'Start date is required']
  },
  endDate: {
    type: Date,
    required: [true, 'End date is required']
  },
  reason: {
    type: String,
    required: [true, 'Reason is required'],
    maxlength: [500, 'Reason cannot exceed 500 characters']
  },
  status: {
    type: String,
    enum: ['pending', 'approved', 'rejected', 'cancelled'],
    default: 'pending'
  },
  appliedDate: {
    type: Date,
    default: Date.now
  },
  approvedBy: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User'
  },
  approvedDate: Date,
  rejectionReason: String,
  emergencyContact: {
    name: String,
    phone: String,
    relation: String
  },
  destination: {
    address: String,
    city: String,
    state: String,
    pincode: String
  },
  attachments: [String],
  actualReturnDate: Date,
  isExtended: {
    type: Boolean,
    default: false
  },
  extensionRequests: [{
    requestedEndDate: Date,
    reason: String,
    requestedDate: Date,
    status: {
      type: String,
      enum: ['pending', 'approved', 'rejected']
    },
    approvedBy: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User'
    },
    approvedDate: Date
  }],
  statusHistory: [{
    status: {
      type: String,
      enum: ['pending', 'approved', 'rejected', 'cancelled']
    },
    updatedBy: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User'
    },
    updatedAt: {
      type: Date,
      default: Date.now
    },
    comment: String
  }],
  parentApproval: {
    required: {
      type: Boolean,
      default: false
    },
    obtained: {
      type: Boolean,
      default: false
    },
    contactNumber: String,
    approvalDate: Date
  },
  medicalCertificate: {
    required: {
      type: Boolean,
      default: false
    },
    uploaded: {
      type: Boolean,
      default: false
    },
    fileName: String,
    uploadDate: Date
  }
}, {
  timestamps: true
});

// Indexes for better query performance
leaveSchema.index({ leaveId: 1 });
leaveSchema.index({ student: 1 });
leaveSchema.index({ status: 1 });
leaveSchema.index({ startDate: 1 });
leaveSchema.index({ endDate: 1 });
leaveSchema.index({ leaveType: 1 });

// Pre-save middleware to generate leave ID
leaveSchema.pre('save', function(next) {
  if (!this.leaveId) {
    const year = new Date().getFullYear();
    const month = String(new Date().getMonth() + 1).padStart(2, '0');
    const random = Math.floor(Math.random() * 10000).toString().padStart(4, '0');
    this.leaveId = `LV${year}${month}${random}`;
  }
  
  // Set requirements based on leave type
  if (this.leaveType === 'medical') {
    this.medicalCertificate.required = true;
  }
  
  // Require parent approval for leaves longer than 7 days
  const leaveDays = Math.ceil((this.endDate - this.startDate) / (1000 * 60 * 60 * 24));
  if (leaveDays > 7) {
    this.parentApproval.required = true;
  }
  
  next();
});

// Validation for date range
leaveSchema.pre('validate', function(next) {
  if (this.startDate && this.endDate && this.startDate >= this.endDate) {
    next(new Error('End date must be after start date'));
  } else {
    next();
  }
});

// Method to update status
leaveSchema.methods.updateStatus = function(newStatus, updatedBy, comment) {
  this.status = newStatus;
  
  this.statusHistory.push({
    status: newStatus,
    updatedBy: updatedBy,
    updatedAt: new Date(),
    comment: comment
  });
  
  if (newStatus === 'approved') {
    this.approvedBy = updatedBy;
    this.approvedDate = new Date();
  } else if (newStatus === 'rejected') {
    this.rejectionReason = comment;
  }
  
  return this.save();
};

// Method to request extension
leaveSchema.methods.requestExtension = function(newEndDate, reason) {
  this.extensionRequests.push({
    requestedEndDate: newEndDate,
    reason: reason,
    requestedDate: new Date(),
    status: 'pending'
  });
  
  return this.save();
};

// Virtual for leave duration in days
leaveSchema.virtual('durationDays').get(function() {
  if (this.startDate && this.endDate) {
    const diffMs = this.endDate - this.startDate;
    return Math.ceil(diffMs / (1000 * 60 * 60 * 24));
  }
  return 0;
});

// Virtual for current status
leaveSchema.virtual('currentStatus').get(function() {
  const now = new Date();
  
  if (this.status !== 'approved') {
    return this.status;
  }
  
  if (now < this.startDate) {
    return 'upcoming';
  } else if (now >= this.startDate && now <= this.endDate) {
    return 'active';
  } else {
    return this.actualReturnDate ? 'completed' : 'overdue';
  }
});

// Virtual for overdue days
leaveSchema.virtual('overdueDays').get(function() {
  if (this.status === 'approved' && !this.actualReturnDate) {
    const now = new Date();
    if (now > this.endDate) {
      const diffMs = now - this.endDate;
      return Math.floor(diffMs / (1000 * 60 * 60 * 24));
    }
  }
  return 0;
});

module.exports = mongoose.model('Leave', leaveSchema);

File: src/main/java/com/hostel/model/Leave.java
package com.hostel.model;

import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.mongodb.core.mapping.Document;

import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Size;
import java.time.LocalDateTime;
import java.time.LocalDate;
import java.time.temporal.ChronoUnit;
import java.util.List;
import java.util.ArrayList;

@Document(collection = "leaves")
public class Leave {
    @Id
    private String id;
    
    private String leaveId;
    
    @NotNull(message = "Student ID is required")
    private String studentId;
    
    private String leaveType; // home, medical, emergency, personal, academic, other
    
    @NotNull(message = "Start date is required")
    private LocalDate startDate;
    
    @NotNull(message = "End date is required")
    private LocalDate endDate;
    
    @Size(max = 500, message = "Reason cannot exceed 500 characters")
    private String reason;
    
    private String status = "PENDING"; // PENDING, APPROVED, REJECTED, CANCELLED
    
    private LocalDateTime appliedDate;
    
    private String approvedBy;
    
    private LocalDateTime approvedDate;
    
    private String rejectionReason;
    
    private EmergencyContact emergencyContact;
    
    private Destination destination;
    
    private List<String> attachments = new ArrayList<>();
    
    private LocalDate actualReturnDate;
    
    private boolean isExtended = false;
    
    private List<ExtensionRequest> extensionRequests = new ArrayList<>();
    
    private List<StatusHistory> statusHistory = new ArrayList<>();
    
    private ParentApproval parentApproval;
    
    private MedicalCertificate medicalCertificate;
    
    @CreatedDate
    private LocalDateTime createdAt;
    
    @LastModifiedDate
    private LocalDateTime updatedAt;
    
    // Constructors
    public Leave() {
        this.appliedDate = LocalDateTime.now();
        generateLeaveId();
    }
    
    public Leave(String studentId, String leaveType, LocalDate startDate, LocalDate endDate, String reason) {
        this();
        this.studentId = studentId;
        this.leaveType = leaveType;
        this.startDate = startDate;
        this.endDate = endDate;
        this.reason = reason;
    }
    
    private void generateLeaveId() {
        String year = String.valueOf(LocalDateTime.now().getYear());
        String month = String.format("%02d", LocalDateTime.now().getMonthValue());
        String random = String.format("%04d", (int)(Math.random() * 10000));
        this.leaveId = "LV" + year + month + random;
    }
    
    // Getters and Setters
    public String getId() { return id; }
    public void setId(String id) { this.id = id; }
    
    public String getLeaveId() { return leaveId; }
    public void setLeaveId(String leaveId) { this.leaveId = leaveId; }
    
    public String getStudentId() { return studentId; }
    public void setStudentId(String studentId) { this.studentId = studentId; }
    
    public String getLeaveType() { return leaveType; }
    public void setLeaveType(String leaveType) { this.leaveType = leaveType; }
    
    public LocalDate getStartDate() { return startDate; }
    public void setStartDate(LocalDate startDate) { this.startDate = startDate; }
    
    public LocalDate getEndDate() { return endDate; }
    public void setEndDate(LocalDate endDate) { this.endDate = endDate; }
    
    public String getReason() { return reason; }
    public void setReason(String reason) { this.reason = reason; }
    
    public String getStatus() { return status; }
    public void setStatus(String status) { this.status = status; }
    
    public LocalDateTime getAppliedDate() { return appliedDate; }
    public void setAppliedDate(LocalDateTime appliedDate) { this.appliedDate = appliedDate; }
    
    public String getApprovedBy() { return approvedBy; }
    public void setApprovedBy(String approvedBy) { this.approvedBy = approvedBy; }
    
    public LocalDateTime getApprovedDate() { return approvedDate; }
    public void setApprovedDate(LocalDateTime approvedDate) { this.approvedDate = approvedDate; }
    
    public String getRejectionReason() { return rejectionReason; }
    public void setRejectionReason(String rejectionReason) { this.rejectionReason = rejectionReason; }
    
    public EmergencyContact getEmergencyContact() { return emergencyContact; }
    public void setEmergencyContact(EmergencyContact emergencyContact) { this.emergencyContact = emergencyContact; }
    
    public Destination getDestination() { return destination; }
    public void setDestination(Destination destination) { this.destination = destination; }
    
    public List<String> getAttachments() { return attachments; }
    public void setAttachments(List<String> attachments) { this.attachments = attachments; }
    
    public LocalDate getActualReturnDate() { return actualReturnDate; }
    public void setActualReturnDate(LocalDate actualReturnDate) { this.actualReturnDate = actualReturnDate; }
    
    public boolean isExtended() { return isExtended; }
    public void setExtended(boolean extended) { isExtended = extended; }
    
    public List<ExtensionRequest> getExtensionRequests() { return extensionRequests; }
    public void setExtensionRequests(List<ExtensionRequest> extensionRequests) { this.extensionRequests = extensionRequests; }
    
    public List<StatusHistory> getStatusHistory() { return statusHistory; }
    public void setStatusHistory(List<StatusHistory> statusHistory) { this.statusHistory = statusHistory; }
    
    public ParentApproval getParentApproval() { return parentApproval; }
    public void setParentApproval(ParentApproval parentApproval) { this.parentApproval = parentApproval; }
    
    public MedicalCertificate getMedicalCertificate() { return medicalCertificate; }
    public void setMedicalCertificate(MedicalCertificate medicalCertificate) { this.medicalCertificate = medicalCertificate; }
    
    public LocalDateTime getCreatedAt() { return createdAt; }
    public void setCreatedAt(LocalDateTime createdAt) { this.createdAt = createdAt; }
    
    public LocalDateTime getUpdatedAt() { return updatedAt; }
    public void setUpdatedAt(LocalDateTime updatedAt) { this.updatedAt = updatedAt; }
    
    public long getDurationDays() {
        return startDate != null && endDate != null ? ChronoUnit.DAYS.between(startDate, endDate) + 1 : 0;
    }
    
    public String getCurrentStatus() {
        LocalDate now = LocalDate.now();
        
        if (!"APPROVED".equals(status)) {
            return status;
        }
        
        if (now.isBefore(startDate)) {
            return "UPCOMING";
        } else if (!now.isAfter(endDate)) {
            return "ACTIVE";
        } else {
            return actualReturnDate != null ? "COMPLETED" : "OVERDUE";
        }
    }
    
    public long getOverdueDays() {
        if ("APPROVED".equals(status) && actualReturnDate == null) {
            LocalDate now = LocalDate.now();
            if (now.isAfter(endDate)) {
                return ChronoUnit.DAYS.between(endDate, now);
            }
        }
        return 0;
    }
    
    // Nested classes
    public static class EmergencyContact {
        private String name;
        private String phone;
        private String relation;
        
        // Getters and Setters
        public String getName() { return name; }
        public void setName(String name) { this.name = name; }
        
        public String getPhone() { return phone; }
        public void setPhone(String phone) { this.phone = phone; }
        
        public String getRelation() { return relation; }
        public void setRelation(String relation) { this.relation = relation; }
    }
    
    public static class Destination {
        private String address;
        private String city;
        private String state;
        private String pincode;
        
        // Getters and Setters
        public String getAddress() { return address; }
        public void setAddress(String address) { this.address = address; }
        
        public String getCity() { return city; }
        public void setCity(String city) { this.city = city; }
        
        public String getState() { return state; }
        public void setState(String state) { this.state = state; }
        
        public String getPincode() { return pincode; }
        public void setPincode(String pincode) { this.pincode = pincode; }
    }
    
    public static class ExtensionRequest {
        private LocalDate requestedEndDate;
        private String reason;
        private LocalDateTime requestedDate;
        private String status = "PENDING";
        private String approvedBy;
        private LocalDateTime approvedDate;
        
        // Getters and Setters
        public LocalDate getRequestedEndDate() { return requestedEndDate; }
        public void setRequestedEndDate(LocalDate requestedEndDate) { this.requestedEndDate = requestedEndDate; }
        
        public String getReason() { return reason; }
        public void setReason(String reason) { this.reason = reason; }
        
        public LocalDateTime getRequestedDate() { return requestedDate; }
        public void setRequestedDate(LocalDateTime requestedDate) { this.requestedDate = requestedDate; }
        
        public String getStatus() { return status; }
        public void setStatus(String status) { this.status = status; }
        
        public String getApprovedBy() { return approvedBy; }
        public void setApprovedBy(String approvedBy) { this.approvedBy = approvedBy; }
        
        public LocalDateTime getApprovedDate() { return approvedDate; }
        public void setApprovedDate(LocalDateTime approvedDate) { this.approvedDate = approvedDate; }
    }
    
    public static class StatusHistory {
        private String status;
        private String updatedBy;
        private LocalDateTime updatedAt;
        private String comment;
        
        public StatusHistory() {}
        
        public StatusHistory(String status, String updatedBy, String comment) {
            this.status = status;
            this.updatedBy = updatedBy;
            this.comment = comment;
            this.updatedAt = LocalDateTime.now();
        }
        
        // Getters and Setters
        public String getStatus() { return status; }
        public void setStatus(String status) { this.status = status; }
        
        public String getUpdatedBy() { return updatedBy; }
        public void setUpdatedBy(String updatedBy) { this.updatedBy = updatedBy; }
        
        public LocalDateTime getUpdatedAt() { return updatedAt; }
        public void setUpdatedAt(LocalDateTime updatedAt) { this.updatedAt = updatedAt; }
        
        public String getComment() { return comment; }
        public void setComment(String comment) { this.comment = comment; }
    }
    
    public static class ParentApproval {
        private boolean required = false;
        private boolean obtained = false;
        private String contactNumber;
        private LocalDateTime approvalDate;
        
        // Getters and Setters
        public boolean isRequired() { return required; }
        public void setRequired(boolean required) { this.required = required; }
        
        public boolean isObtained() { return obtained; }
        public void setObtained(boolean obtained) { this.obtained = obtained; }
        
        public String getContactNumber() { return contactNumber; }
        public void setContactNumber(String contactNumber) { this.contactNumber = contactNumber; }
        
        public LocalDateTime getApprovalDate() { return approvalDate; }
        public void setApprovalDate(LocalDateTime approvalDate) { this.approvalDate = approvalDate; }
    }
    
    public static class MedicalCertificate {
        private boolean required = false;
        private boolean uploaded = false;
        private String fileName;
        private LocalDateTime uploadDate;
        
        // Getters and Setters
        public boolean isRequired() { return required; }
        public void setRequired(boolean required) { this.required = required; }
        
        public boolean isUploaded() { return uploaded; }
        public void setUploaded(boolean uploaded) { this.uploaded = uploaded; }
        
        public String getFileName() { return fileName; }
        public void setFileName(String fileName) { this.fileName = fileName; }
        
        public LocalDateTime getUploadDate() { return uploadDate; }
        public void setUploadDate(LocalDateTime uploadDate) { this.uploadDate = uploadDate; }
    }
}